### PR TITLE
Other invoices made inaccessible for customer to download

### PR DESCRIPTION
**Pull request by bartoszpietrzak1994**

### CODE REVIEW

File: spec/Generator/InvoicePdfFileGeneratorSpec.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $this->shouldHaveType(InvoicePdfFileGenerator::class);
+    }
+
+    function it_implements_invoice_pdf_file_generator_interface(): void
+    {
+        $this->shouldImplement(InvoicePdfFileGeneratorInterface::class);
+    }
+
+    function it_creates_credit_memo_pdf_with_generated_content_and_filename_basing_on_credit_memo_number(
```
Comment (Zales0123) [inline spec/Generator/InvoicePdfFileGeneratorSpec.php]: `creates_credit` -> `invoice` :)

File: spec/ResponseBuilder/InvoiceFileResponseBuilderSpec.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+use Symfony\Component\HttpFoundation\Response;
+
+final class InvoiceFileResponseBuilderSpec extends ObjectBehavior
+{
+    function it_is_initializable(): void
+    {
+        $this->shouldHaveType(InvoiceFileResponseBuilder::class);
+    }
+
+    function it_implements_credit_memo_file_response_builder_interface(): void
```
Comment (Zales0123) [inline spec/ResponseBuilder/InvoiceFileResponseBuilderSpec.php]: `credit_memo_file_response_builder_interface` -> `invoice_file_response_builder_interface`

**Comment by Zales0123** [comment]
Would be nice to have it fixed :) also rebase is required

File: src/Ui/Action/Shop/DownloadInvoiceAction.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        $this->invoiceFileResponseBuilder = $invoiceFileResponseBuilder;
+        $this->customerContext = $customerContext;
+    }
+
+    public function __invoke(string $id): Response
+    {
+        /** @var CustomerInterface $customer */
+        $customer = $this->customerContext->getCustomer();
+
+        $invoicePdfFile = null;
```
Comment (Zales0123) [inline src/Ui/Action/Shop/DownloadInvoiceAction.php]: I think it's not needed

File: spec/Checker/InvoiceCustomerRelationCheckerSpec.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        $invoice->orderNumber()->willReturn('00002');
+
+        $orderRepository->findOneByNumber('00002')->willReturn($order);
+
+        $order->getCustomer()->willReturn($customer);
+
+        $customer->getId()->willReturn(1);
+
+        $this->check('00001', $customer);
```
Comment (Zales0123) [inline spec/Checker/InvoiceCustomerRelationCheckerSpec.php]: I think something has gone wrong with this spec :D We have only one mock for 2 customers (from parameter and from order)

File: spec/Checker/InvoiceCustomerRelationCheckerSpec.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    ): void {
+        $this->beConstructedWith($invoiceRepository, $orderRepository);
+    }
+
+    function it_implements_invoice_customer_relation_checker_interface(): void
+    {
+        $this->shouldImplement(InvoiceCustomerRelationCheckerInterface::class);
+    }
+
+    function it_checks_if_customer_id_from_order_is_equal_to_customer_id_from_customer_context(
```
Comment (Zales0123) [inline spec/Checker/InvoiceCustomerRelationCheckerSpec.php]: `it_checks_if_customer_id_from_order_is_equal_to_passed_customer_id`? We're not using context here :)

File: spec/Generator/InvoicePdfFileGeneratorSpec.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        $this->beConstructedWith(
+            $invoiceRepository,
+            $channelRepository,
+            $twig,
+            $pdfGenerator,
+            'invoiceTemplate.html.twig'
+        );
+    }
+
+    function it_is_initializable(): void
```
Comment (Zales0123) [inline spec/Generator/InvoicePdfFileGeneratorSpec.php]: We have this method in some specs, in some we don't #consistencyFTW

File: spec/Invoice/Admin/AdminInvoiceDownloaderSpec.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    function it_implements_admin_invoice_downloader_interface(): void
+    {
+        $this->shouldImplement(AdminInvoiceDownloaderInterface::class);
+    }
+
+    function it_returns_pdf_file_with_an_invoice(InvoicePdfFileGeneratorInterface $invoicePdfFileGenerator): void
+    {
+        $invoicePdf = new InvoicePdf('', '');
```
Comment (Zales0123) [inline spec/Invoice/Admin/AdminInvoiceDownloaderSpec.php]: Can't we put some values here? Empty strings looks awful :(

File: src/Checker/InvoiceCustomerRelationChecker.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+final class InvoiceCustomerRelationChecker implements InvoiceCustomerRelationCheckerInterface
+{
+    /** @var InvoiceRepository */
+    private $invoiceRepository;
+
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    public function __construct(
+        InvoiceRepository $invoiceRepository,
```
Comment (pamil) [inline src/Checker/InvoiceCustomerRelationChecker.php]: `InvoiceRepositoryInterface`?

File: src/Checker/InvoiceCustomerRelationCheckerInterface.php [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\InvoicingPlugin\Checker;
+
+use Sylius\Component\Core\Model\CustomerInterface;
+
+interface InvoiceCustomerRelationCheckerInterface
+{
+    public function check(string $invoiceId, CustomerInterface $customer): void;
```
Comment (pamil) [inline src/Checker/InvoiceCustomerRelationCheckerInterface.php]: I'm missing the docblock about the exception thrown when a customer can't access the given invoice.

Comment (pamil) [inline src/Checker/InvoiceCustomerRelationCheckerInterface.php]: Also, I would rename the method to actually say what does it check (canAccess)?

File: src/Email/InvoiceEmailSender.php [hunk 10]
```diff
@@ -4,67 +4,46 @@
...
-
-        $pdfFileContent = $this->pdfGenerator->getOutputFromHtml(
-            $this->templatingEngine->render('@SyliusInvoicingPlugin/Resources/views/Invoice/Download/pdf.html.twig', [
-                'invoice' => $invoice,
-                'channel' => $channel,
-        ]));
+        $invoicePdfFile = $this->invoicePdfFileGenerator->generate($invoice->id());
 
-        $filePath = $this->preparePdfFilePath('invoice-%s.pdf', $invoice->id());
+        $fileName = $invoicePdfFile->filename();
```
Comment (pamil) [inline src/Email/InvoiceEmailSender.php]: Is it `file name` or `filename`? 🤔

File: src/Invoice/Admin/AdminInvoiceDownloaderInterface.php [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\InvoicingPlugin\Invoice\Admin;
+
+use Sylius\InvoicingPlugin\Model\InvoicePdf;
+
+interface AdminInvoiceDownloaderInterface
+{
+    public function download(string $invoiceId): InvoicePdf;
```
Comment (pamil) [inline src/Invoice/Admin/AdminInvoiceDownloaderInterface.php]: Does it download an invoice? It looks like a provider / repository.

File: features/managing_invoices/being_unable_to_download_other_users_invoice.feature [hunk 12]
```diff
@@ -0,0 +1,20 @@
+@customer_browsing_invoices @application
```
Comment (pamil) [inline features/managing_invoices/being_unable_to_download_other_users_invoice.feature]: Could we implement this on UI level also?

**Comment by pamil** [comment]
Thank you, Bartosz! :tada:

**Pull request merged**